orderer/common/filerepo/filerepo.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package filerepo
+
+import (
+	"bufio"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/hyperledger/fabric/common/configtx"
+)
+
+type Repo struct {
+	Options Options
+}
+
+type Options struct {
+	FileSuffix          string
+	TransientFilePrefix string
+	FileRepoDir         string
+}
+
+// Save atomically persists the content to the channelName+suffix file by first writing it
+// to a tmp file prefixed by the TransientFilePrefix and then moves the file to the final
+// destination indicated by the FileSuffix.
+func (r *Repo) Save(channelName string, content []byte) error {
+	err := configtx.ValidateChannelID(channelName)
+	if err != nil {
+		return err
+	}
+
+	fileRepoDir := r.Options.FileRepoDir
+	dest := filepath.Join(fileRepoDir, channelName+r.Options.FileSuffix)
+
+	tmpFile, err := ioutil.TempFile(fileRepoDir, r.Options.TransientFilePrefix+channelName)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err != nil {
+			tmpFile.Close()
+			os.Remove(tmpFile.Name())
+		}
+	}()
+
+	w := bufio.NewWriter(tmpFile)
+
+	if _, err := w.Write(content); err != nil {
+		return err
+	}
+
+	if err := w.Flush(); err != nil {
+		return err
+	}
+
+	if err := tmpFile.Close(); err != nil {
+		return err
+	}
+
+	return os.Rename(tmpFile.Name(), dest)
+}
+
+// Scan parses the directory and produce a list of file names, filtered by suffix.
+func (r *Repo) Scan() ([]string, error) {
+	var repoFiles []string
+
+	files, err := ioutil.ReadDir(r.Options.FileRepoDir)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, f := range files {
+		isFileSuffix, err := filepath.Match("*"+r.Options.FileSuffix, f.Name())
+		if err != nil {
+			return nil, err
+		}
+		if isFileSuffix {
+			repoFiles = append(repoFiles, f.Name())
+		}
+	}
+
+	return repoFiles, nil
+}
+
+// Load reads the file contents.
+func (r *Repo) Load(fileName string) ([]byte, error) {
```

Comment:
while the `Save()` receives a channelName, shouldn't load also receive the channel name? Maybe I am asking this because I don't have clarity on how the consumer would use it.

Reply:
I thought about this initially but after thinking about it more I'm not sure if we should prefer taking a channel name or file name inherently. The `filerepo` could ideally be used for more general purpose applications outside of channel participation related usages; however taking a file name in for the `Save` function would require changing some of the logic around the temp file path. Maybe for now for standardization purposes we can use the `filerepo` exclusively for channel participation and modify `Load` to receive the channel name as well.